src/tactic/library_search.lean

```diff
@@ -0,0 +1,105 @@
+-- Copyright (c) 2019 Scott Morrison. All rights reserved.
+-- Released under Apache 2.0 license as described in the file LICENSE.
+-- Authors: Scott Morrison
+
+import tactic.basic
+import data.list.defs
+import data.option.basic
+
+/-
+A basic `library_search` tactic.
+
+There's more to come:
+* using `↔` lemmas
+* full-blown backwards reasoning with search, using multiple library results...
+-/
+
+namespace tactic
+open native
+
+namespace library_search
+
+-- TODO clean up here
+meta def head_symbol : expr → name
```

Comment:
Instead of matching on `n` for `gt` or `ge`, you could use `whnf` on some `app` nodes.

Reply:
I'm not really sure what to do here. I don't want to run any tactics while traversing the library, to keep that fast.

Reply:
I understand. The other thing you could do is to create an attribute with a cache to indicate definitions that must be used as synonyms for each other. `reducible` might be usable in this way.

Reply:
@semorrison Could a preprocessing step help? That would mean that `library_search` looks at some dump that isn't `.lean`. If the dump is binary, it should probably not be tracked by `git`, which is annoying. But otherwise that might be able to give good speedups, right?

Reply:
@cipher1024 for now I think I'd like to "wait and see", and see if people find many cases where the current hack results in `library_search` failing to find an obvious solution. I don't want to build a piece of infrastructure here that isn't actually needed. Of course I appreciate that the current `head_symbol` is a hack, however. :-(

Reply:
@jcommelin, it seems that traversing the library is not actually that expensive. Parsing files is almost certainly going to be worse than what happens now.

I worry that the current bottleneck is actually in my `iff` processing, where I call `infer_type` on every lemma that we try. This isn't necessary, as the relevant type is actually directly available (without going via tactics) from the `declaration`. It will take a little plumbing to get that type to the point of use, but I think I know how to do it.